File: src/engine/gameEngine.ts
import type { GameState } from '../types/game'
import type { Player, PlayerId } from '../types/player'
import type { Board } from '../types/board'
import type { MovementOrder, AllOrders } from '../types/orders'
import type { PlayerStats, EndGameStats } from '../types/stats'
import type { TurnStep } from './turnResolver'
import { generateBoard } from './boardGenerator'
import { generateUnits } from './unitGenerator'
import { resolvePlayerTurn } from './turnResolver'
import { PLAYER_COLORS, PLAYER_NAMES } from '../types/player'
import { computeAiOrders } from '../ai/aiController'

const HUMAN_PLAYER_ID: PlayerId = 'p0'
const AI_COUNT = 5

function createInitialStats(playerIds: PlayerId[]): Map<PlayerId, PlayerStats> {
  return new Map(playerIds.map(id => [id, {
    playerId: id,
    unitsGenerated: 0,
    unitsKilled: 0,
    tilesConquered: 0,
    tilesLost: 0,
    tilesAtEnd: 0,
  }]))
}

export function initGame(): GameState {
  const playerIds: PlayerId[] = Array.from({ length: 1 + AI_COUNT }, (_, i) => `p${i}`)
  const players: Player[] = playerIds.map((id, i) => ({
    id,
    type: i === 0 ? 'human' : 'ai',
    color: PLAYER_COLORS[i],
    name: PLAYER_NAMES[i],
    isEliminated: false,
  }))

  const board = generateBoard(playerIds)
  const orders: AllOrders = new Map(playerIds.map(id => [id, new Map()]))
  const runningStats = createInitialStats(playerIds)

  return {
    phase: 'playerTurn',
    board,
    players,
    humanPlayerId: HUMAN_PLAYER_ID,
    orders,
    humanStandingOrders: new Map(),
    turn: { turnNumber: 1, activeAiIndex: 0 },
    winner: null,
    stats: null,
    runningStats,
  }
}

export function setHumanStandingOrder(state: GameState, order: MovementOrder): GameState {
  const newStanding = new Map(state.humanStandingOrders)
  newStanding.set(order.fromKey, order)
  return { ...state, humanStandingOrders: newStanding }
}

export function cancelHumanStandingOrder(state: GameState, fromKey: string): GameState {
  const newStanding = new Map(state.humanStandingOrders)
  newStanding.delete(fromKey)
  return { ...state, humanStandingOrders: newStanding }
}

function cleanupStandingOrders(standing: OrderMap, board: Board, humanPlayerId: PlayerId): OrderMap {
  const cleaned = new Map(standing)
  for (const [fromKey] of cleaned) {
    const tile = board.get(fromKey)
    if (!tile || tile.owner !== humanPlayerId) cleaned.delete(fromKey)
  }
  return cleaned
}

function applyStandingOrdersToFreshOrders(
  freshOrders: AllOrders,
  standing: OrderMap,
  board: Board,
  humanPlayerId: PlayerId,
): AllOrders {
  const humanOrders = new Map<string, MovementOrder>()
  for (const [fromKey, order] of standing) {
    const tile = board.get(fromKey)
    if (tile && tile.owner === humanPlayerId && tile.units > 0) {
      humanOrders.set(fromKey, order)
    }
  }
  const result = new Map(freshOrders)
  result.set(humanPlayerId, humanOrders)
  return result
}

export function applyHumanOrder(state: GameState, order: MovementOrder): GameState {
  const humanOrders = new Map(state.orders.get(state.humanPlayerId) ?? new Map())
  humanOrders.set(order.fromKey, order)
  const newOrders = new Map(state.orders)
  newOrders.set(state.humanPlayerId, humanOrders)
  return { ...state, orders: newOrders }
}

export function cancelHumanOrder(state: GameState, fromKey: string): GameState {
  const humanOrders = new Map(state.orders.get(state.humanPlayerId) ?? new Map())
  humanOrders.delete(fromKey)
  const newOrders = new Map(state.orders)
  newOrders.set(state.humanPlayerId, humanOrders)
  return { ...state, orders: newOrders }
}

export function executeHumanMoves(state: GameState): {
  newState: GameState
  steps: TurnStep[]
} {
  const humanOrders = state.orders.get(state.humanPlayerId) ?? new Map()
  const result = resolvePlayerTurn(
    state.board,
    state.players,
    humanOrders,
    state.humanPlayerId,
    new Map(state.runningStats),
  )

  if (result.winnerId) {
    return buildEndState(state, result.board, result.players, result.runningStats, result.winnerId, result.steps)
  }

  const humanPlayer = result.players.find(p => p.id === state.humanPlayerId)
  if (!humanPlayer || humanPlayer.isEliminated) {
    return buildEndState(state, result.board, result.players, result.runningStats, null, result.steps)
  }

  const cleanedStanding = cleanupStandingOrders(state.humanStandingOrders, result.board, state.humanPlayerId)
  return {
    newState: {
      ...state,
      board: result.board,
      players: result.players,
      runningStats: result.runningStats,
      humanStandingOrders: cleanedStanding,
      phase: 'playerTurn',
    },
    steps: result.steps,
  }
}

export function endHumanTurn(state: GameState): GameState {
  return {
    ...state,
    phase: 'aiTurn',
    turn: { ...state.turn, activeAiIndex: 0 },
  }
}

export function resolveAiTurn(state: GameState, aiIndex: number): {
  newState: GameState
  steps: TurnStep[]
} {
  const aiPlayers = state.players.filter(p => p.type === 'ai' && !p.isEliminated)
  if (aiIndex >= aiPlayers.length) {
    // All AIs done — generate units and start next player turn
    const updatedStats = new Map(state.runningStats)
    const newBoard = generateUnits(state.board, updatedStats)
    const cleanedStanding = cleanupStandingOrders(state.humanStandingOrders, newBoard, state.humanPlayerId)
    const freshOrders: AllOrders = new Map(state.players.map(p => [p.id, new Map()]))
    const ordersWithStanding = applyStandingOrdersToFreshOrders(freshOrders, cleanedStanding, newBoard, state.humanPlayerId)
    const nextState: GameState = {
      ...state,
      board: newBoard,
      orders: ordersWithStanding,
      humanStandingOrders: cleanedStanding,
      runningStats: updatedStats,
      phase: 'playerTurn',
      turn: { turnNumber: state.turn.turnNumber + 1, activeAiIndex: 0 },
    }
    return { newState: nextState, steps: [] }
  }

  const aiPlayer = aiPlayers[aiIndex]
  const aiOrders = state.orders.get(aiPlayer.id) ?? new Map()
  const freshAiOrders = computeAiOrders(state.board, aiPlayer.id, aiOrders, state.players)

  const newOrdersMap = new Map(state.orders)
  newOrdersMap.set(aiPlayer.id, freshAiOrders)

  const result = resolvePlayerTurn(
    state.board,
    state.players,
    freshAiOrders,
    aiPlayer.id,
    new Map(state.runningStats),
  )

  if (result.winnerId) {
    return buildEndState(state, result.board, result.players, result.runningStats, result.winnerId, result.steps)
  }

  const humanPlayer = result.players.find(p => p.id === state.humanPlayerId)
  if (!humanPlayer || humanPlayer.isEliminated) {
    return buildEndState(state, result.board, result.players, result.runningStats, null, result.steps)
  }

  return {
    newState: {
      ...state,
      board: result.board,
      players: result.players,
      runningStats: result.runningStats,
      orders: newOrdersMap,
      phase: 'aiTurn',
      turn: { ...state.turn, activeAiIndex: aiIndex + 1 },
    },
    steps: result.steps,
  }
}

function buildEndState(
  state: GameState,
  board: Board,
  players: Player[],
  runningStats: Map<PlayerId, PlayerStats>,
  winnerId: PlayerId | null,
  steps: TurnStep[],
): { newState: GameState; steps: TurnStep[] } {
  const outcome = winnerId === state.humanPlayerId ? 'win' : 'lose'

  const finalStats: EndGameStats = {
    outcome,
    winnerId,
    playerStats: players.map(p => {
      const s = runningStats.get(p.id) ?? { playerId: p.id, unitsGenerated: 0, unitsKilled: 0, tilesConquered: 0, tilesLost: 0, tilesAtEnd: 0 }
      const tilesAtEnd = [...board.values()].filter(t => t.owner === p.id).length
      return { ...s, tilesAtEnd }
    }),
  }

  return {
    newState: {
      ...state,
      board,
      players,
      runningStats,
      phase: 'end',
      winner: winnerId,
      stats: finalStats,
    },
    steps,
  }
}

export function retireGame(state: GameState): GameState {
  const finalStats: EndGameStats = {
    outcome: 'retire',
    winnerId: null,
    playerStats: state.players.map(p => {
      const s = state.runningStats.get(p.id) ?? { playerId: p.id, unitsGenerated: 0, unitsKilled: 0, tilesConquered: 0, tilesLost: 0, tilesAtEnd: 0 }
      const tilesAtEnd = [...state.board.values()].filter(t => t.owner === p.id).length
      return { ...s, tilesAtEnd }
    }),
  }
  return { ...state, phase: 'end', stats: finalStats }
}

File: research/hexwar/engine/game_engine.py
"""
Headless HexWar game engine for AI training and evaluation.

This module provides HexWarEnv: a self-contained simulation that runs a
complete 6-player game to completion. It mirrors the JS gameEngine.ts logic
but exposes a clean Python interface suited for:
  - Rollout collection (AI training)
  - Tournament evaluation
  - Game replay and analysis

Turn structure (mirrors JS):
  1. Each player (in order p0..p5) submits orders and resolves their turn.
  2. After all players move, generate_units() runs for every tile.
  3. Repeat until win condition or max_turns.

The JS game separates "human turn" and "AI turn" phases for the UI. Here,
all players are treated identically — the caller provides an agent callable
for each player.
"""

from __future__ import annotations

from random import Random
from typing import Callable

from .board_generator import generate_board
from .turn_resolver import TurnResult, resolve_player_turn
from .types import (
    PLAYER_COLORS,
    PLAYER_IDS,
    PLAYER_NAMES,
    Board,
    GameResult,
    GameState,
    OrderMap,
    Player,
    PlayerStats,
    PlayerType,
    TurnState,
)
from .unit_generator import generate_units

# ---------------------------------------------------------------------------
# Agent protocol
# ---------------------------------------------------------------------------

# An agent callable receives (board, player_id, players, stats) and returns
# an OrderMap (fromKey → MovementOrder). Return an empty dict to pass.
AgentFn = Callable[[Board, str, list[Player], dict[str, PlayerStats]], OrderMap]


# ---------------------------------------------------------------------------
# Environment
# ---------------------------------------------------------------------------

MAX_TURNS_DEFAULT = 300


class HexWarEnv:
    """
    Headless HexWar game simulation.

    Usage::

        from hexwar.engine.game_engine import HexWarEnv
        from hexwar.agents.random_agent import RandomAgent

        agents = {pid: RandomAgent() for pid in PLAYER_IDS}
        env = HexWarEnv(agents=agents, seed=42)
        result = env.run()
        print(result.winner_id, result.turns_played)

    Attributes:
        board:    Current board state.
        players:  Current player list (including eliminated).
        stats:    Per-player running stats.
        turn:     Current turn state.
        phase:    "playerTurn" | "generateUnits" | "end"
    """

    def __init__(
        self,
        agents: dict[str, AgentFn] | None = None,
        player_ids: list[str] | None = None,
        seed: int | None = None,
        rng: Random | None = None,
        max_turns: int = MAX_TURNS_DEFAULT,
    ) -> None:
        """
        Initialise a new game.

        Args:
            agents:     Dict mapping player_id → AgentFn. Players without
                        an agent pass each turn (no orders).
            player_ids: Override the default 6-player list.
            seed:       RNG seed for reproducible board generation.
            rng:        Pre-constructed Random instance (overrides seed).
            max_turns:  Hard cap; game ends as draw if exceeded.
        """
        self.player_ids = player_ids or PLAYER_IDS
        self.agents: dict[str, AgentFn] = agents or {}
        self.max_turns = max_turns
        self._elimination_order: list[str] = []

        # Generate board
        _rng = rng if rng is not None else Random(seed)
        self.board: Board = generate_board(
            player_ids=self.player_ids,
            rng=_rng,
        )

        # Build players
        self.players: list[Player] = [
            Player(
                id=pid,
                name=PLAYER_NAMES[i] if i < len(PLAYER_NAMES) else pid,
                color=PLAYER_COLORS[i] if i < len(PLAYER_COLORS) else "#ffffff",
                type=PlayerType.AI,
            )
            for i, pid in enumerate(self.player_ids)
        ]

        # Initialise stats
        self.stats: dict[str, PlayerStats] = {
            pid: PlayerStats(player_id=pid)
            for pid in self.player_ids
        }

        self.turn = TurnState(turn_number=1, active_ai_index=0)
        self.phase = "playerTurn"
        self.winner_id: str | None = None
        # Fixed turn order: always iterate over the same player list in order.
        # active_ai_index indexes into self.player_ids (not just live players)
        # so eliminations mid-round don't cause skipping.

    # ------------------------------------------------------------------
    # Main run loop
    # ------------------------------------------------------------------

    def run(self) -> GameResult:
        """
        Run the game to completion and return the result.

        Returns:
            GameResult with winner, turns played, final stats, and
            elimination order.
        """
        while self.phase != "end":
            self.step()
        return GameResult(
            winner_id=self.winner_id,
            turns_played=self.turn.turn_number - 1,
            stats=self.stats,
            elimination_order=self._elimination_order,
        )

    def step(self) -> None:
        """
        Advance the game by one logical step.

        One step = one player's turn OR the unit-generation phase.
        Callers that want fine-grained control can call step() in a loop
        instead of run().
        """
        if self.phase == "end":
            return

        if self.phase == "playerTurn":
            self._resolve_next_player()
        elif self.phase == "generateUnits":
            self._do_generate_units()

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _resolve_next_player(self) -> None:
        """Resolve the current player's turn, then advance to the next.

        We iterate over player_ids in fixed order (not just live players) so
        that mid-round eliminations don't cause indices to shift and skip turns.
        """
        idx = self.turn.active_ai_index
        if idx >= len(self.player_ids):
            # All players have had their turn — transition to unit generation
            self.phase = "generateUnits"
            return

        pid = self.player_ids[idx]
        player = next((p for p in self.players if p.id == pid), None)

        # Advance index regardless — the player may be eliminated or missing
        self.turn.active_ai_index += 1

        if player is None or player.is_eliminated:
            return  # Skip eliminated players silently
        agent = self.agents.get(player.id)
        orders: OrderMap = {}
        if agent is not None:
            orders = agent(self.board, player.id, self.players, self.stats)

        result: TurnResult = resolve_player_turn(
            board=self.board,
            players=self.players,
            orders=orders,
            player_id=player.id,
            stats=self.stats,
        )

        self.board = result.board
        self.players = result.players
        self.stats = result.stats

        # Track newly eliminated
        for p in self.players:
            if p.is_eliminated and p.id not in self._elimination_order:
                self._elimination_order.append(p.id)

        if result.winner_id:
            self._end_game(winner_id=result.winner_id)

    def _do_generate_units(self) -> None:
        """Run the unit generation phase and start the next turn."""
        generate_units(self.board, self.stats)

        if self.turn.turn_number >= self.max_turns:
            self._end_game(winner_id=None)
            return

        self.turn = TurnState(
            turn_number=self.turn.turn_number + 1,
            active_ai_index=0,
        )
        self.phase = "playerTurn"

    def _end_game(self, winner_id: str | None) -> None:
        self.winner_id = winner_id
        self.phase = "end"

    # ------------------------------------------------------------------
    # Observation helpers (used by AI agents)
    # ------------------------------------------------------------------

    def get_game_state(self) -> GameState:
        """
        Return a GameState snapshot of the current game.

        This is a convenience method for agents that expect a GameState.
        The returned state is a shallow copy — do not mutate.
        """
        return GameState(
            board=self.board,
            players=self.players,
            human_player_id=self.player_ids[0],
            orders={},
            human_standing_orders={},
            phase=self.phase,
            turn=self.turn,
            stats=self.stats,
        )
